### ISSUE TITLE
Menu entries in Windows/high contrast not always readable.

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/Pale-Moon#1591]
This is a related issue to #137 but with a different cause (the effect is not the same).
Since it's reported on Windows 10, it may not apply to all windows versions either. Some more information/testing is required.

@tomasz1986 can you please detail exactly what you have in use? Which browser theme is in use (and is it reproducible on the default theme?), which high contrast system theme are you using, and does it persist if you change the option in preferences -> content -> colors -> `override the colors...` to `Never`? 

![1nut6eg](https://user-images.githubusercontent.com/3359132/35146195-8bfb7e7c-fd0a-11e7-94cd-30098fb14849.png)

**Comment by tomasz1986** [comment]
This is my own customized theme but I have just checked the stock ones, and the exact same issue takes place with the default High Contrast White theme. On the other hand, it does not happen with High Contrast Black (or is rather simply not visible, as the theme's default background colour is also black).

It actually seems that this specific highlight color is hardcoded to black in the browser. Below is a screen shot when using also a custom theme, but this time based on High Contrast Black. Everything is readable, but you can still see that the background colour has changed to black after moving the mouse over the menu entries. The text is only readable because the actual font colour is originally white in the theme.

![palemoon_highcontrastblack](https://user-images.githubusercontent.com/5626656/35149848-8b4a087c-fd0f-11e7-8af2-f1b343cee661.png)

In short, the problem for sure occurs with High Contrast White based themes. The `override the colors...` setting makes no difference here.

**Comment by mattatobin** [comment]
With enough conrtast you CAN see it.

**Comment by wolfbeast** [comment]
Thanks for the additional info. This is most likely a widget code problem (C++), and something that does **not** occur on Windows 7. I'll be looking into this, as I'm pretty sure I know where this goes wrong.

**Comment by tomasz1986** [comment]
Thank you for such a quick reaction. The menus are still usable in the current state, although navigating through bookmarks is kind of difficult (which is also one of the reasons why I switched to a bookmark sidebar).

>With enough conrtast you CAN see it.
Yeah, I can still see the text in the second case (black Menu Bar letters), but in case of the highlight issue, it is straight black on black.

**Comment by mattatobin** [comment]
No i meant with enough contrast you can ses it ISN'T black on black.

**Comment by tomasz1986** [comment]
> No i meant with enough contrast you can ses it ISN'T black on black.

Are you really speaking about the first screen shot linked to by @wolfbeast? It is black on black... I have even checked with a colour picker in Paint.NET. It comes up as straight rgb(0,0,0).

**Comment by wolfbeast** [comment]
I'm building a new unstable right now with a potential fix in. should be done and published in an hour or so -- could you do me a favor and let me know if this unstable version helps with your issue?

**Comment by tomasz1986** [comment]
Thank you. I have just tested the unstable build. The highlight issue seems to be gone :)

![palemoon_highcontrastwhite](https://user-images.githubusercontent.com/5626656/35171921-5e2bf35c-fd5d-11e7-879b-20efd82ba077.png)

There are still some graphic glitches with the dividing lines though. Nothing serious or urgent to fix, but looks weird. It is visible in the screen shot above. Short bold lines are added to the left of the existing ones, and their position is also not aligned correctly.

As for high contrast black based themes, you can check the following screen shot.

![palemoon_highcontrastblack](https://user-images.githubusercontent.com/5626656/35172024-ed16ee28-fd5d-11e7-9aa9-4519ab86e7dc.png)

The "extended" lines are also visible here, and the Menu Bar font is still black.

I do understand that your fix addressed only the highlight problem. I am just writing for information with further possible fixes in mind.

**Comment by wolfbeast** [comment]
Thank you for testing and confirming; looks like I was on the right track.

You have to understand that high contrast themes are not just alternative system themes; they are special high-accessibility themes that make some very specific changes to how the entire desktop compositor, windows and their controls, and window metrics work. Many people use them to get a "dark system theme" but they don't understand the other changes it makes can seriously break things or require special case workarounds in software to deal with.

The lines you'll just have to live with. The main menu I can have a look at as a follow-up here, but that's about as far as this can be special-cased.

**Comment by tomasz1986** [comment]
Yes, I do understand that.

However, I also do in fact need these kind of high contrast themes in Windows 10 though. Not the dark ones, but rather the light ones. I find the overall darkish Windows 10 default theme very difficult to use (especially the light text and icons on a black taskbar). In the previous versions of Windows it was possible to adjust the theme colours enough, so I did not need to enable any high contrast themes to adjust the OS theme to my needs. In Windows 10, the taskbar is bound to use only dark colours with white text on top of it. I find it very difficult for my eyes :(

The black theme displayed above I may only use sometimes at night when there is very little light in the room, as otherwise the monitor is simply too bright to look at.

**Comment by wolfbeast** [comment]
The current unstable is as far as this styling can go; it should solve at least some of the main menu readability issues, although not all of them.

**Comment by wolfbeast** [comment]
The transparency optimization bypass should be uplifted to Basilisk.

**Comment by tomasz1986** [comment]
I have been busy in the last few days and until today had no time to check the browser after the latest changes.

The Menu Bar now changes text on hover - from white to black with white high contrast themes, and from black to white with black high contrast themes. I would say that it is better than before, although the upper Menu text is still black and black with black themes when selected.

![1](https://user-images.githubusercontent.com/5626656/35255672-7e4dda78-0033-11e8-8246-930bdf56bb57.png)

I guess that there is no simple method to just keep it stay the same colour all the time? My custom CSS that sets the menu text colour to `initial` does that, but I am not familiar with the browser internals to know if such a change is possible (or makes sense) here.

**Comment by wolfbeast** [comment]
Nothing else left to do here. Closing.

**Comment by tomasz1986** [comment]
I am writing again since Pale Moon 28 beta 4 still has this issue (while Pale Moon 27.9.4 does not). Is the fix planned to be applied later in the development cycle, or is it a different problem?

**Comment by wolfbeast** [comment]
Fixed in UXP
